File: foundational_design_patterns/11_structured_outputs/src/structured_outputs_advanced.py
"""Structured Outputs (advanced): Instructor retry + malformed input failure mode."""

from __future__ import annotations

import os
import sys

from pathlib import Path

ROOT_DIR = next(parent for parent in Path(__file__).resolve().parents if (parent / "ssl_fix.py").exists())
if str(ROOT_DIR) not in sys.path:
    sys.path.insert(0, str(ROOT_DIR))

from repo_support import configure_example, get_advanced_model

configure_example(__file__)

from pydantic import BaseModel, Field, ValidationError


class ExtractedInvoice(BaseModel):
    vendor: str
    invoice_id: str
    total_amount_usd: float = Field(..., ge=0)
    due_date: str


BROKEN_INVOICE = """
Supplier maybe: ???
Invoice id unclear maybe INV-X
Total: TBD
Due: soon-ish
"""


def instructor_extract(text: str) -> ExtractedInvoice:
    import instructor
    from openai import OpenAI

    client = instructor.from_openai(OpenAI())
    return client.chat.completions.create(
        model=get_advanced_model(),
        response_model=ExtractedInvoice,
        messages=[
            {"role": "system", "content": "Extract invoice fields. If missing, infer conservatively but validly."},
            {"role": "user", "content": text},
        ],
        max_retries=2,
    )


def main() -> None:
    print("\n" + "=" * 80)
    print("STRUCTURED OUTPUTS — ADVANCED (INSTRUCTOR + FAILURE MODE)")
    print("=" * 80)

    if not os.getenv("OPENAI_API_KEY"):
        print("❌ OPENAI_API_KEY is required.")
        return

    try:
        extracted = instructor_extract(BROKEN_INVOICE)
        print("\nInstructor extraction output:")
        print(extracted.model_dump())
    except ValidationError as exc:
        print("\nValidation failed even after retries (expected failure mode):")
        print(exc)
    except Exception as exc:
        print("\nExtraction error:")
        print(exc)

    print("\nNote: For self-hosted models, grammar-constrained decoding (Outlines/XGrammar) is recommended.")


if __name__ == "__main__":
    main()

File: foundational_design_patterns/11_structured_outputs/typescript/src/structured_outputs_advanced.ts
/**
 * Structured Outputs (advanced) — TypeScript port of src/structured_outputs_advanced.py.
 */

import OpenAI from "openai";
import { generateObject } from "ai";
import { openai } from "@ai-sdk/openai";
import { z } from "zod";

export const ExtractedInvoiceSchema = z.object({
  vendor: z.string(),
  invoice_id: z.string(),
  total_amount_usd: z.number().min(0),
  due_date: z.string(),
});

export type ExtractedInvoice = z.infer<typeof ExtractedInvoiceSchema>;

export const BROKEN_INVOICE = `
Supplier maybe: ???
Invoice id unclear maybe INV-X
Total: TBD
Due: soon-ish
`;

export async function instructorExtract(
  text: string,
): Promise<ExtractedInvoice> {
  const model = process.env.OPENAI_ADVANCED_MODEL ?? process.env.OPENAI_MODEL ?? "gpt-5.2";
  let lastError: unknown;

  // Mirrors max_retries=2 in spirit: initial try + 2 retries.
  for (let attempt = 0; attempt < 3; attempt++) {
    try {
      const result = await generateObject({
        model: openai(model),
        schema: ExtractedInvoiceSchema,
        system:
          "Extract invoice fields. If missing, infer conservatively but validly.",
        prompt: text,
      });
      return result.object;
    } catch (error) {
      lastError = error;
    }
  }

  throw lastError;
}

async function main(): Promise<void> {
  console.log("\n" + "=".repeat(80));
  console.log("STRUCTURED OUTPUTS — ADVANCED (INSTRUCTOR + FAILURE MODE)");
  console.log("=".repeat(80));

  if (!process.env.OPENAI_API_KEY) {
    console.log("❌ OPENAI_API_KEY is required.");
    return;
  }

  try {
    const extracted = await instructorExtract(BROKEN_INVOICE);
    console.log("\nInstructor extraction output:");
    console.log(extracted);
  } catch (exc: unknown) {
    console.log("\nValidation failed even after retries (expected failure mode):");
    console.log(exc);
  }

  console.log(
    "\nNote: For self-hosted models, grammar-constrained decoding (Outlines/XGrammar) is recommended.",
  );
}

if (import.meta.main) {
  // Touch OpenAI client import path so the runtime mirrors the Python example's
  // explicit OpenAI/instructor dependency surface.
  void OpenAI;
  await main();
}
